torch/utils/tensorboard/_pytorch_graph.py

```diff
@@ -0,0 +1,267 @@
+
+import torch
+from ._proto_graph import node_proto
+from torch.onnx.utils import OperatorExportTypes
+
+
+methods_OP = ['attributeNames', 'hasMultipleOutputs', 'hasUses', 'inputs',
+              'kind', 'outputs', 'outputsSize', 'scopeName']
+methods_IO = ['node', 'offset', 'uniqueName']  # 'unique' <int> , 'type' <Tensor<class 'torch._C.Type'>>
+
+
+class NodeBase(object):
+    def __init__(self, uniqueName=None, inputs=None, scope=None, tensor_size=None, op_type='UnSpecified', attributes=''):
+        self.uniqueName = uniqueName
+        self.inputs = inputs
+        self.tensor_size = tensor_size
+        self.kind = op_type
+        self.attributes = attributes
+        if scope is not None:
+            self.scope = scope
+
+    def __repr__(self):
+        repr = []
+        repr.append(str(type(self)))
+        for m in dir(self):
+            if '__' not in m:
+                repr.append(m + ': ' + str(getattr(self, m)) + str(type(getattr(self, m))))
+        return '\n'.join(repr) + '\n\n'
+
+
+class NodePy(NodeBase):
+    def __init__(self, NodeCpp, valid_methods):
+        super(NodePy, self).__init__(NodeCpp)
+        self.valid_methods = valid_methods[:]
+        self.inputs = []
+
+        for m in self.valid_methods:
+            if m == 'inputs' or m == 'outputs':
+                list_of_node = list(getattr(NodeCpp, m)())
+                io_unique_names = []
+                io_tensor_sizes = []
+                for n in list_of_node:
+                    io_unique_names.append(n.uniqueName())
+                    if n.type().kind() == 'CompleteTensorType':
+                        io_tensor_sizes.append(n.type().sizes())
+                    else:
+                        io_tensor_sizes.append(None)
+
+                setattr(self, m, io_unique_names)
+                setattr(self, m + 'tensor_size', io_tensor_sizes)
+
+            else:
+                setattr(self, m, getattr(NodeCpp, m)())
+
+
+class NodePyIO(NodePy):
+    def __init__(self, NodeCpp, input_or_output=None):
+        super(NodePyIO, self).__init__(NodeCpp, methods_IO)
+        try:
+            tensor_size = NodeCpp.type().sizes()
+        except RuntimeError:
+            tensor_size = [1, ]  # fail when constant model is used.
+        self.tensor_size = tensor_size
+        self.kind = 'Parameter'
+        if input_or_output:
+            self.input_or_output = input_or_output
+            self.kind = 'IO Node'
+
+
+class NodePyOP(NodePy):
+    def __init__(self, NodeCpp):
+        super(NodePyOP, self).__init__(NodeCpp, methods_OP)
+        self.attributes = str({k: NodeCpp[k] for k in NodeCpp.attributeNames()}).replace("'", ' ')
+        self.kind = NodeCpp.kind()
+
+
+class GraphPy(object):
+    def __init__(self):
+        self.nodes_op = []
+        self.nodes_io = OrderedDict()
+        self.uniqueNameToScopedName = {}
+        self.shallowestScopeName = 'default'
+        self.scope_name_appeared = []
```

Comment:
It would be very helpful if there were explanations of what these attributes hold.

Reply:
@lanpa can you comment here? Thanks.

Reply:
The GraphDef generation is a two-pass operation. At first, all nodes are read and save to two lists. One list is for Input/Output nodes, who don't have both in-bound and out-bound connections. Another list is for nodes_OP (operation). The first-pass also saves all scope name appeared in the nodes in `scope_name_appeared` list for later processing.

second pass:
`uniqueNameToScopedName` is a mapping from a node's ID to its fully qualified scope name. e.g. `Net1/Linear[0]/1`. It will be trivial to implement if pytorch.jit have totally correct scope output. The function `populate_namespace_from_OP_to_IO` and `findCommonRoot` are used to assign scope name to a node based on the connection between nodes. (heuristic way, again)

Reply:
`shallowestScopeName` and `scope_name_appeared` is used to book-keeping in second pass.

Reply:
Yeah, so what I'd like, is what you just wrote here, to be put in the code, as actual comments. (This comment does not require action for experimental merge.)

Reply:
Added a task for this.